File: Recursion/subsequence_pattern/combination_sum_1.cpp
#include <iostream>
using namespace std;
void helper(int ind, vector<int> &arr, int target, vector<vector<int>> &ans, vector<int> &ds)
{
    if (target == 0)
    {
        ans.push_back(ds);
        return;
    }
    for (int i = ind; i < arr.size(); i++)
    {
        if (target < arr[i])
            break; // as the array is sorted so the next number might be bigger than this
        ds.push_back(arr[i]);
        helper(i, arr, target - arr[i], ans, ds);
        ds.pop_back();
    }
}
vector<vector<int>> combinationSum(vector<int> &arr, int target)
{
    sort(arr.begin(), arr.end());
    vector<vector<int>> ans;
    vector<int> ds;
    helper(0, arr, target, ans, ds);
    return ans;
}

File: Recursion/subsequence_pattern/CombinationSum1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSum1 {
    public static List<List<Integer>> combinationSum(int[] arr, int target) {
        Arrays.sort(arr);
        List<List<Integer>> ans = new ArrayList<>();
        helper(0, arr, target, ans, new ArrayList<>());
        return ans;
    }

    private static void helper(int idx, int[] arr, int target, List<List<Integer>> ans, List<Integer> ds) {
        if (target == 0) {
            ans.add(new ArrayList<>(ds));
            return;
        }

        for (int i = idx; i < arr.length; i++) {
            if (arr[i] > target) {
                break;
            }
            ds.add(arr[i]);
            helper(i, arr, target - arr[i], ans, ds);
            ds.remove(ds.size() - 1);
        }
    }
}